libyul/YulName.cpp

```diff
@@ -0,0 +1,429 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/YulName.h>
+
+#include <libyul/backends/evm/EVMDialect.h>
+#include <libyul/optimiser/NameCollector.h>
+#include <libyul/AST.h>
+#include <libyul/Dialect.h>
+#include <libyul/Exceptions.h>
+
+#include <fmt/compile.h>
+
+namespace solidity::yul
+{
+
+YulNameRepository::YulNameRepository(solidity::yul::Dialect const& _dialect):
+	m_dialect(_dialect)
+{
+	m_predefined.empty = defineName("");
+	yulAssert(m_predefined.empty == YulNameRepository::emptyName());
+
+	for (auto const& type: _dialect.types)
+		if (type.empty())
+		{
+			m_indexBoundaries.beginTypes = 0;
+			m_dialectTypes.emplace_back(emptyName(), type.str());
+		}
+		else
+		{
+			m_indexBoundaries.beginTypes = 1;
+			m_dialectTypes.emplace_back(defineName(type.str()), type.str());
+		}
+	m_indexBoundaries.endTypes = m_index;
+	m_indexBoundaries.beginBuiltins = m_index;
+
+	auto const& builtinNames = _dialect.builtinNames();
+	m_predefined.verbatim = defineName("@ verbatim");
+	for (auto const& label: builtinNames)
+		if (!label.empty())
+		{
+			auto const name = defineName(label);
+			if (auto const* function = m_dialect.get().builtin(YulString(label)))
+				m_builtinFunctions[name] = convertBuiltinFunction(name, *function);
+		}
+	m_indexBoundaries.endBuiltins = m_index;
+
+	m_predefined.boolType = nameOfType(_dialect.boolType.str());
+	m_predefined.defaultType = nameOfType(_dialect.defaultType.str());
+
+	auto const predefinedName = [&](std::string const& label)
+	{
+		if (builtinNames.count(label) > 0)
+			return nameOfBuiltin(label);
+		else
+			return defineName(label);
+	};
```

Comment:
Wouldn't it make more sense to set this to `emptyName()` when a builtin isn't available?

For example I see that #15215 will use this for `memoryguard()` in `EVMObjectCompiler`. `EVMObjectCompiler` currently runs `FunctionCallFinder` to find `memoryguard()` calls. The change would require adding an extra check there, but that should probably be an assert anyway, because I think the assumption there is that we're using a dialect in which it is a builtin. If it weren't, we could get a false positive by finding a random user-defined function that happens to be called `memoryguard()`.

Reply:
Overall, I'm not sure this `PredefinedHandles` thing is a good idea, because it inherently makes assumptions about what builtins are available. It feels like we should just accept whatever is available in `_dialect.builtinNames()` without making too many assumptions and expose it via some accessor (that also allows 0for things to be missing).

Also, more generally, it seems odd to me that `YulName` even depends on `Dialect` and not the other way around it. But I still haven't read all the PRs so I don't have a complete picture. I suspect that there was a good reason for doing it this way.

Reply:
Yes, I think `emptyName()` would be better.

> Overall, I'm not sure this PredefinedHandles thing is a good idea, because it inherently makes assumptions about what builtins are available. It feels like we should just accept whatever is available in _dialect.builtinNames() without making too many assumptions and expose it via some accessor (that also allows for things to be missing).

I added them so that the `YulNameRepository` can act as a sort of proxy for a `Dialect` and builtins (or other special handles) then should have efficient access. Which kind of ties into your previous point but also into your following point:

> Also, more generally, it seems odd to me that YulName even depends on Dialect and not the other way around it.

One of my issues with the `YulString`s was the repository's static lifetime and (in my eyes) odd cleanup callbacks once in a while. So I tried changing that into a proper instance with lifetime. Now `Dialect`s are also managed with static lifetime but are immutable once construction is completed, so this directly clashes with (hypothetically) doing it the other way around because the `YulNameRepository` instances are highly mutable and also shouldn't be static, I think. That's why I settled for wrapping the dialect. Now I have stared at this thing for so long that it seems natural to me to do it this way around, I would be curious to hear why you would see the dependency rather inverted :)